### PR TITLE
Retries

### PR DESCRIPTION
**Pull request by bertie-sektorcert**
Following summary is AI generated (but read and verified). Do you have an opinion @unode?

## Summary

Implements the retry behavior discussed in https://github.com/embl-bio-it/python-mattermost-autodriver/issues/42: the driver now automatically retries requests that fail due to server-side rate limiting or transient errors, following the conventions of established SDKs (slack_sdk, botocore, the OpenAI/Anthropic clients) rather than pulling in a retry library.

## Retry policy

| Failure | Retried? | Notes |
|---|---|---|
| HTTP 429 | All methods | Safe for any method — the server rejected the request outright. Honors `Retry-After` / `X-RateLimit-Reset`. |
| Connection errors, 502/503/504 | Idempotent methods only (GET/PUT/DELETE/HEAD) | A POST may already have been processed; retrying risks duplicated messages. |
| Requests with `files` or streaming `data=` bodies | Never | Stream bodies are consumed on first send; a replay would transmit an empty body. |
| 401 and other 4xx | Never | Deterministic failures; retrying cannot help. |

Two new driver options control this: `max_retries` (default 3, `0` disables) and `retry_max_sleep` (default 30 s, caps any single wait — if the server requests a longer wait the request fails immediately instead of blocking).

## New `TooManyRequests` exception

Mattermost's rate limiter replies with a plain-text body rather than the standard JSON error ([`ratelimit.go`](https://github.com/mattermost/mattermost/blob/master/server/channels/app/ratelimit.go)), so 429s previously surfaced as `InvalidMattermostError` or `UnknownMattermostError` and could not be caught cleanly. 429 detection now happens from the status code and headers before body parsing, raising `TooManyRequests` with a `retry_after` attribute. Wait values are parsed in all three conventions found in the wild: delay-seconds, HTTP-date, and absolute unix timestamps.

**Backwards incompatible:** code catching `UnknownMattermostError` / `InvalidMattermostError` for 429s must switch to `TooManyRequests` (noted in the changelog).

## Tests and CI

The PR adds the project's first test suite: 46 pytest tests covering the existing status→exception mapping and the new retry behavior for both `Client` and `AsyncClient`, using `httpx.MockTransport` (no new runtime dependencies — dev dependencies live in `requirements-dev.txt`). A new workflow runs the suite on pull requests across Python 3.10–3.13.

## Notes for review

- Retry logic is hand-rolled (~80 lines shared in `BaseClient`) rather than a Tenacity dependency — the decision needs the response object (status, headers, original method), and sync/async variants mirror the existing client split.
- Websocket reconnection is untouched; token-refresh-on-401 is deliberately out of scope (it is a re-authentication concern, not a retry concern).

### CODE REVIEW

File: src/mattermostautodriver/client.py [hunk 1]
```diff
@@ -166,21 +179,142 @@ def filter_dict_or_none(d):
...
+        """Parse a rate limit header value into seconds to wait from now.
+
+        Handles the conventions used in the wild:
+
+        - delay in seconds ("120") - RFC 7231 ``Retry-After`` and Mattermost's
+          relative ``X-RateLimit-Reset``
+        - HTTP-date ("Wed, 21 Oct 2015 07:28:00 GMT") - the alternative
+          ``Retry-After`` form, sent by some proxies and CDNs
+        - absolute unix timestamp ("1794000000") - the ``X-RateLimit-Reset``
+          convention of many API gateways
```
Comment (unode) [inline src/mattermostautodriver/client.py:197]: According to [Mattermost docs](https://developers.mattermost.com/api-documentation/#/#rate-limiting) they implement the `epoch`-based interface with `X-RateLimit-Reset`.
The `seconds` and `date` options are only mentioned in [RFC 7231 but in the context of `Retry-After`](https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.3).

Instead of handling the 3 cases as the same and then "guess" if we have an epoch timestamp, we probably want to stick closer to the two specifications and handle the two headers separately:
- Use the `seconds` or `date` approaches for `Retry-After`
- Use `epoch` for `X-RateLimit-Reset`
(hoping to remove the `EPOCH_THRESHOLD` logic above which looks hackish.)

I was also going to suggest that we use an ISO 8601 date but seems like the field follows the format of [RFC 5322](https://datatracker.ietf.org/doc/html/rfc5322#section-3.3).

File: src/mattermostautodriver/client.py [hunk 2]
```diff
@@ -236,10 +371,24 @@ def make_request(self, method, endpoint, options=None, params=None, data=None, f
...
+                if delay is None:
+                    raise
+                log.warning("Received %r - retrying in %.1f seconds", e, delay)
+            else:
+                delay = self._retry_delay(method, attempt, data=data, files=files, response=response)
+                if delay is None:
+                    self._check_response(response)
+                    return response
+                log.warning("Received status %d - retrying in %.1f seconds", response.status_code, delay)
+            time.sleep(delay)
```
Comment (unode) [inline src/mattermostautodriver/client.py:397]: With the current `backoff`, `sleep` may cause extended blocking which, depending on the case, may result in unexpected reply "lag" or delays.
The new `retry_max_sleep` argument addresses this somewhat but `retries * 30s` can still feel like an eternity.

We may want to highlight this side-effect in the documentation and recommend the `AsyncClient` interface if users hit this and it becomes a problem for them.

File: tests/test_async_client.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    handler, calls = sequence_handler([rate_limit_response({"Retry-After": "2"}), httpx.Response(200, json={"ok": 1})])
+    client = make_async_client(handler, max_retries=3)
+
+    assert await client.get("/users/me") == {"ok": 1}
+    assert len(calls) == 2
+    assert sleeps == [2.0]
+
+
+async def test_429_raises_after_retries_are_exhausted(sleeps):
+    handler, calls = sequence_handler([rate_limit_response({"Retry-After": "0"})])
```
Comment (unode) [inline tests/test_async_client.py:34]: Took me a moment to understand this test. Adding a comment for clarity. Feel free to reword.

```suggestion
    # Without an ok response in `sequence_handler`, this test will repeat "Retry-After 0 seconds" for all requests, exhausting the client `max_retries`.
    handler, calls = sequence_handler([rate_limit_response({"Retry-After": "0"})])
```

File: tests/test_client.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    assert BaseClient._parse_wait_time("120") == 120.0
+
+
+def test_parse_wait_time_clamps_negative_to_zero():
+    assert BaseClient._parse_wait_time("-5") == 0.0
+
+
+def test_parse_wait_time_http_date():
+    value = format_datetime(datetime.now(timezone.utc) + timedelta(seconds=60))
+    assert 55 < BaseClient._parse_wait_time(value) <= 60
```
Comment (unode) [inline tests/test_client.py]: Can we document the magic `55` here and in the following tests? Maybe make it a global to this test file?
I understand it to be an arbitrary, and rather generous, lower boundary.

In practice we should see these tests always return something between 59 and 60 as the function should run in less than 1 second. Right?

File: tests/test_client.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+
+def test_429_with_epoch_ratelimit_reset_is_retried(sleeps):
+    headers = {"X-RateLimit-Reset": str(int(time.time()) + 5)}
+    handler, calls = sequence_handler([rate_limit_response(headers), httpx.Response(200, json={"ok": 1})])
+    client = make_client(handler, max_retries=3)
+
+    assert client.get("/users/me") == {"ok": 1}
+    assert len(calls) == 2
+    assert 0 < sleeps[0] <= 6
```
Comment (unode) [inline tests/test_client.py]: Same consideration here? Do we really expect tests to sleep anywhere between 0 and 6 seconds?

Also wondering, should we be type-strict and assert with a `float` rather than `int`? (same for other tests below)

File: tests/test_client.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    client = make_client(lambda request: error_response(429))
+
+    with pytest.raises(TooManyRequests) as excinfo:
+        client.get("/users/me")
+
+    assert excinfo.value.error_id == "api.some.error"
+    assert excinfo.value.request_id == "req1234"
+
+
+def test_post_sends_json_options():
```
Comment (unode) [inline tests/test_client.py]: This test seems a little out of place or out of order.

Could you add some comments explaining the motivation for this test?
If we are testing to ensure our POST requests to Mattermost are in JSON format, should we perhaps try to parse the request and make sure it is not only JSON but also a valid request given the inputs?

File: tests/test_client.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    assert len(calls) == 2
+    assert sleeps == [2.0]
+
+
+def test_429_is_retried_for_post(sleeps):
+    handler, calls = sequence_handler([rate_limit_response({"Retry-After": "0"}), httpx.Response(200, json={"ok": 1})])
+    client = make_client(handler, max_retries=3)
+
+    assert client.post("/posts", options={"message": "hi"}) == {"ok": 1}
+    assert len(calls) == 2
```
Comment (unode) [inline tests/test_client.py:194]: Should we have this for completion?


```suggestion
    assert len(calls) == 2
    assert sleeps == [0.0]
```

File: tests/test_client.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+    handler, calls = sequence_handler([rate_limit_response({"Retry-After": "2"}), httpx.Response(200, json={"ok": 1})])
+    client = make_client(handler, max_retries=3)
+
+    assert client.get("/users/me") == {"ok": 1}
+    assert len(calls) == 2
+    assert sleeps == [2.0]
+
+
+def test_429_is_retried_for_post(sleeps):
+    handler, calls = sequence_handler([rate_limit_response({"Retry-After": "0"}), httpx.Response(200, json={"ok": 1})])
```
Comment (unode) [inline tests/test_client.py:190]: Should this test be expanded to include the cases where POST is **not** to be retried such as those with streams or consumed bodies?
I see these are tested below but not specifically for 429 (unless I missed it).

File: tests/test_client.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+    handler, calls = sequence_handler([rate_limit_response({"Retry-After": "0"})])
+    client = make_client(handler, max_retries=0)
+
+    with pytest.raises(TooManyRequests):
+        client.get("/users/me")
+
+    assert len(calls) == 1
+
+
+def test_retry_after_above_cap_raises_immediately(sleeps):
```
Comment (unode) [inline tests/test_client.py:230]: Nice, hadn't thought of this edge case until here!

File: tests/test_client.py [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+    handler, calls = sequence_handler([rate_limit_response({"Retry-After": "0"})])
+    client = make_client(handler, max_retries=3)
+
+    with pytest.raises(TooManyRequests):
+        client.post("/files", files={"files": b"content"})
+
+    assert len(calls) == 1
+
+
+@pytest.mark.filterwarnings("ignore::DeprecationWarning")  # httpx warns about stream bodies via data=
```
Comment (unode) [inline tests/test_client.py:290]: To be fixed? Should we open an issue if we are using a deprecated interface for streams?

File: src/mattermostautodriver/client.py [hunk 11]
```diff
@@ -236,10 +371,24 @@ def make_request(self, method, endpoint, options=None, params=None, data=None, f
...
+        while True:
+            try:
+                response = request(url + endpoint, **request_params)
+            except httpx.TransportError as e:
+                delay = self._retry_delay(method, attempt, data=data, files=files)
+                if delay is None:
+                    raise
+                log.warning("Received %r - retrying in %.1f seconds", e, delay)
+            else:
+                delay = self._retry_delay(method, attempt, data=data, files=files, response=response)
```
Comment (unode) [inline src/mattermostautodriver/client.py:392]: Could we add a comment here indicating which cases (HTTP codes / Failures) are handled by the `TransportError` block and which we rely on headers to decide if we should retry?

**Review by unode — APPROVED** [review]
Hi @bertie-sektorcert 

Added some comments but they are mostly aesthetic and covering some reflections while I was reading and following the code mentally.

Overall I think this looks good and I was nitpicking.

I have a few high-level comments/questions:

1. Since we have two client interfaces `Client` and `AsyncClient`, should we perhaps have the same test coverage in both of them? The tests in this PR are a but unbalanced towards the synchronous client, which is understandable but I wondered if we should keep them identical to catch unexpected differences.
2. The documentation doesn't currently mention that `Retry-After` is given preference over `X-RateLimit-Reset`. Worth mentioning since this is something we also explicitly test?
3. Tests now also check different exceptions for different HTTP codes. Would you consider adding a table with "HTTP codes + Exception + retry behavior" in the docs? Might be a useful addition.

**Comment by bertie-sektorcert** [comment]
Thank you once again for your rather thorough review - it's always a pleasure receiving those! A couple of comments on your high level remarks:

1. The tests here were done mostly as a best-effort, and as such, they don't entirely live up to my standards for the rest of the codebase. Personally, I don't think your point about the magic number 55 matters when it's located in simple unit tests, but I do however see your point about not having test parity between the sync and async client. I will look into having these covered symmetrically.
2. Good point - I will add that to the docs
3. Good point too - I will add that too.

Regarding thoughts on X-Ratelimit-* headers vs Retry-After:

AFAIK, Retry-After is the standard (which is why you found it in the RFC) - the X-Ratelimit headers are to my knowledge industry conventions popularized by Twitter and Github. This is also the reason the code prioritizes Retry-After. The intentions of parsing both was originally that while Mattermost only returns one of them, it wouldn't be unrealistic that a load-balancer or other infrastructure placed in front might return another one.

However, it does appear that Mattermost may return both headers. While they document the X-Ratelimit style, `setRateLimitHeaders` in the Mattermost source code (`server/channels/app/ratelimit.go`) actually adds Retry-After alongside the 3 documented X-Ratelimit headers. Therefore, I have settled on prioritizing what is the actual standard, and adding the X-Ratelimit as a sensible fallback.

I did consider whether I should split up the parsing for the two different approaches. I ended up not doing it, because there is no realistic way in which the different data types could overlap. If anything, splitting it up, and not doing the hacky epoch check may give issues. Given that the X-Ratelimit is more of a convention than a standard, it wouldn't seem unreasonable to think that some project somewhere maybe would return seconds till retry refresh instead of the agreed upon epoch. The current "hacky" approach would cover this case.

File: src/mattermostautodriver/client.py (hunk 1 above)
Comment (bertie-sektorcert) [reply src/mattermostautodriver/client.py:197]: Please refer to my thoughts in the comment below

File: src/mattermostautodriver/client.py (hunk 2 above)
Comment (bertie-sektorcert) [reply src/mattermostautodriver/client.py:397]: Yup - just added that to the docs

File: tests/test_client.py (hunk 4 above)
Comment (bertie-sektorcert) [reply tests/test_client.py]: I have marked that - I don't think there's any point to tighten it below 5 seconds though

File: tests/test_client.py (hunk 6 above)
Comment (bertie-sektorcert) [reply tests/test_client.py]: Good point - this kind of testing would be useful, but for now, request encoding/generation testing is out of scoped. I have removed it

File: tests/test_client.py (hunk 10 above)
Comment (bertie-sektorcert) [reply tests/test_client.py:290]: No, actually, this test is designed to test for an edge case that can only occur if the user "misuses" the libary. We never pass stream bodies via data (that is deprecated after all) - but it is possible for the user to do so (and still technically supported by httpx for the time being). As such we test for whether the retry integration handles that correctly here.

File: tests/test_client.py (hunk 8 above)
Comment (bertie-sektorcert) [reply tests/test_client.py:190]: Added

File: src/mattermostautodriver/client.py (hunk 11 above)
Comment (bertie-sektorcert) [reply src/mattermostautodriver/client.py:392]: Good catch - added

**Comment by unode** [comment]
Thanks for the replies. I'm still a bit split by the two types of rate headers but your reasoning is ok so feel free to merge.
